### DISCUSSION TITLE
Accessing octree to recreate the slice but for 3D

### DISCUSSION THREAD

**Discussion by MihirDharmadhikari** [ethz-asl/wavemap#86] [Q&A]
Hello,

Thanks for the great work!
I see that in the map visualization, you create the slice which is split into squares of different resolutions. I wanted to ask two questions about the same:
1. Is there a way to get the same for 3D? 
2. Can we query the occupancy (or probability) of a point at a level of the octree such that the occupancy is returned as occupied if any one of the max resolution voxels that this voxel encompasses are occupied, otherwise return free?

Best,
Mihir

**Comment by victorreijgwart** [comment]
Hi Mihir,  

It's nice to hear from you!  

To answer your first question:  
You can use the `forEachLeaf` method to visit all the (multi-resolution) leaf nodes in the map. We use the same approach to create the slice visualization (code available [here](https://github.com/ethz-asl/wavemap/blob/main/interfaces/ros1/wavemap_rviz_plugin/src/visuals/slice_visual.cc#L103-L140)). In the slice visualization, nodes are skipped if they don't intersect with the plane at the slice's height. By simply removing [that condition](https://github.com/ethz-asl/wavemap/blob/main/interfaces/ros1/wavemap_rviz_plugin/src/visuals/slice_visual.cc#L105-L108), you could operate in 3D.  

For your second question:  
Yes, it's possible to query occupancy in the way you described. There are actually a few different ways you could do it. If you don't mind describing your intended application, I can recommend specific options. Some specific questions I have are:  
- How performance-sensitive is your use case? Are you querying the occupancy only once (e.g., for evaluation) or at a high frequency (e.g., during planning)?  
- Are you working with a static map, or will you be querying a map that's still changing or being built?  
- Do you need to classify the leaves for the entire map or just specific regions?  
- Which API would you prefer to use (C++, Python, or ROS1)?  

Best wishes,  
Victor

**Reply by MihirDharmadhikari** [reply]
Hi Victor,

It's nice to hear from you as well! I hope everything is going well. 

Thanks for pointing to the code about the slice, I will have a look. Is it possible to get the blocks only in a certain cuboid (axis aligned) in the `forEachLeaf` function?

Regarding the second question:
- I plan to using for path planning. I want to query the voxels at different resolution for fast collision-checking. The frequency of this will not be very high, say once every 5 seconds.
- The map won't be static, it will be updated as new measurements come in. Think of it like the collision checking that GBPlanner would need to do. 
- I don't need the leaves of the entire map. I will be querying them for a given cuboid (potentially axis aligned). 
- I would use the C++ API.

Let me know what you think.

Best regards,
Mihir

**Reply by victorreijgwart** [reply]
For anyone following this thread, Mihir and I continued our discussion over chat. If you have similar questions, feel free to reply here or start a new discussion - I’d be glad to answer.
